<font face='Liberation serif' size=3>
    
#### **Привет!** 👋

Меня зовут Артем Пиляй. Я буду проверять твой проект. В рамках корпоративной этики мы общаемся на «ты». Но если лучше на «Вы», то нужно обозначить это сразу.
 
В тетрадке есть комментарии к проекту, они будут в отдельных ячейках Markdown.
 
<div class="alert alert-danger">

**Артём:** ❌ Такой комментарий **нужно исправить** обязательно, он критически влияет на удачное выполнение проекта.

</div>

<div class="alert alert-success">

**Артём:** 👍 Это **главный комментарий**, значит все сделано хорошо. 😉 А еще в таких комментариях бывают подсказки как сделать **твой проект ещё лучше**.

</div>

---

**Распространенные ошибки**

<div class="alert alert-danger">

**Артём:** ❌ Не перемещай и не удаляй мои комментарии. Лучше, чтобы твои ответы на мои комментарии шли по очереди, как сообщения в мессенджере. Это ускорит проверку.
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Проверь проект на работоспособность перед отправкой на проверку. Это важный навык, который тебе пригодится как на собеседовании, при выполнении тестового задания, так и перед совещанием у руководства компании. Помни, ревьеры не вносят изменения в код, даже если это путь к датасету или ячейка в неверном формате (code/mardown).  
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Кстати, смотри как можно сделать, чтобы датасет грузился и локально, и в JupyterHub:

```py
try:
    df = pd.read_csv('C:/Users/Student/Desktop/dataset.csv')
except:
    df = pd.read_csv('/datasets/dataset.csv')
```
    
Используя эту конструкцию в будущих проектах, ты будешь спокойно отправлять проект на проверку зная, что он будет работать.
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Проект должен быть структурирован. Где названия разделов и подразделов будут записаны в ячейках [Markdown](https://paulradzkov.com/2014/markdown_cheatsheet/?).
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Если с твоим проектом будет знакомиться другой человек, например рекрутер или старший дата специалист, в нем должны быть описаны поставленные задачи, содержание датасета, промежуточные и финальные выводы. Как правило, заказчик читает только финальный вывод, поэтому важно добавить информацию о проведенном исследовании, применить форматирование текста, в общем, сделать красиво, чтобы было понятно, что проведено настоящее исследование и выводу можно верить. 

</div>  

<div class="alert alert-success">

**Артём:** 👍 Длинный код и комментарии нужно переносить на новую строку. Согласись, коллеги не оценят, если им придется тратить время на пролистывание ячеек. Согласно международному стандарту [PEP8](https://www.python.org/dev/peps/pep-0008/) длина строки не должна превышать 79 символов.
Вот как можно перенести код на новую строку:

```py
# Переносить можно код внутри скобки
example = (2 *
           2)
```
Конечно, я не считаю сколько символов в строке, главное, чтобы код или комментарий не ушел за пределы ячейки (зависит от разрешения экрана). Если ушел, то это желтый комментарий для каждой нечитаемой ячейки.
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Тимлид, старший дата специалист или ревьюер ценят твое и своё время, подсвечивай свои комментарии в ячейках Markdown. Шаблон я тебе оставил ниже. Подробнее про синтаксис Markdown, с примерами, можно почитать [здесь](https://paulradzkov.com/2014/markdown_cheatsheet/?).
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Буду рад, если воспользуешься моими советами. Они тебе помогут при трудоустройстве, например в [Яндексе](https://yandex.ru/jobs/pages/analytics_interview?ysclid=l84s63fs1g790141490), и при работе в команде. Если в ходе проведения исследования возникнут вопросы, не стесняйся задавай. 🙂
</div>

<div class="alert alert-info">

**Артём:** 🏁 [Итоги ревью здесь](#results)
</div>


**Поехали!** 🚀 👨‍🚀
<a id="start"></a>

<div style="background-color:lavender; color:#black; text-align:left; padding:2%">
    
#### **Студент(ка):** ✅ Это шаблон для тебя. Твой комментарий или вопрос в такой ячейке я точнно увижу.

# Кредитный рейтинг

## Оглавление

[1 Загрузка данных](#1-загрузка-данных)

[2 Оценка полноты данных](#2-оценка-полноты-данных)

[3. Замена пропусков](#3-замена-пропусков)

[4. Приведение типов](#4-приведение-типов)

[5. Обработка дубликатов](#5-обработка-дубликатов)

[6. Категоризация данных ](#6-категоризация-данных)

[7. Проверка гипетез](#7-проверка-гипетез)

[8. Вывод](#8-вывод)

## 1. Загрузка данных

Импортируем модули и загружаем датасет. Знакомимся с данными на примере первых 20 строк датасета.

In [2]:
import pandas as pd
from matplotlib import pyplot as plt

try:
    data = pd.read_csv(r"C:\Users\yka\Documents\Practicum\DS_Practicum\Credit_Scoring\source\data.csv")
except:
    data = pd.read_csv("/datasets/data.csv")

data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Знакомимся с методанными датасета для приведения к верным типам данных.

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


# 2. Оценка полноты данных

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В данных есть пропуски. Подозрительно то, что пропуски только в двух столбцах и их одинаковое количество.

In [5]:
len(data[(data["days_employed"].isna()) & (data["total_income"].isna())])

2174

Люди, у которых нет информации о стаже работы не имеют также и информации о совокупных доходах. Рассмотрим подробнее корреляции между строками, в которых отсутствует информация о стаже и доходах.

In [6]:
data[data["days_employed"].isna()].corr()

children  days_employed  dob_years  education_id  \
children          1.000000            NaN  -0.152431     -0.033552   
days_employed          NaN            NaN        NaN           NaN   
dob_years        -0.152431            NaN   1.000000      0.091549   
education_id     -0.033552            NaN   0.091549      1.000000   
family_status_id -0.036475            NaN  -0.066621     -0.016376   
debt             -0.005732            NaN  -0.052709      0.054495   
total_income           NaN            NaN        NaN           NaN   

                  family_status_id      debt  total_income  
children                 -0.036475 -0.005732           NaN  
days_employed                  NaN       NaN           NaN  
dob_years                -0.066621 -0.052709           NaN  
education_id             -0.016376  0.054495           NaN  
family_status_id          1.000000  0.005102           NaN  
debt                      0.005102  1.000000           NaN  
total_income                   NaN       NaN           NaN

Проверка связи между колонками с пропусками в колонках `days_employed` и `total_income` показала, что корелляция пропусков в этой колонке с другими атрибутами данных низкая. Следовательно, можно сделать вывод, что пропуски допущены случайно.

# 3. Замена пропусков

Пропуски в `total_income` заменим пропущенными медианными значениями для соответствующих пропущенным столбцам типов занятости.

In [7]:
median_income_by_type = data.groupby("income_type")["total_income"].median()

for type_key in median_income_by_type.keys():
    data.loc[(data["total_income"].isna()) & (data["income_type"] == type_key), "total_income"] = median_income_by_type[type_key]
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

Нормализуем данные в `days_employed`.

In [8]:
data["days_employed"] = abs(data["days_employed"])
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Рассмотрим для каждого типа занятости медианное значение трудового стажа `days_employed` в днях.

In [9]:
days_employed_by_type = data.groupby("income_type")["days_employed"].median()
days_employed_by_type

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

Изучим перечень уникальных значений столбца `children`.

In [10]:
data["children"].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

Удалим в столбце children аномальные значения -1 и 20 (будем считать, что это выброс).

In [11]:
data = data.drop(data[(data["children"] == -1) | (data["children"] == 20)].index)
data["children"].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

Т.к. пропуски в столбце `days_employed` случайны, заполним пропуски в столбце `days_employed` медианными значениями по каждому типу занятости `income_type` для сохранения более типичных значений для каждого из классов заёмщиков.

In [12]:
days_employed_by_type = data.groupby("income_type")["days_employed"].median()

for income_type in days_employed_by_type.keys():
    data.loc[(data["days_employed"].isna()) & (data["income_type"] == income_type), "days_employed"] = days_employed_by_type[income_type]

data["days_employed"].isna().sum()

0

# 4. Приведение типов

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.

In [13]:
data["total_income"] = data["total_income"].astype("int")
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21402 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21402 non-null  int64  
 1   days_employed     21402 non-null  float64
 2   dob_years         21402 non-null  int64  
 3   education         21402 non-null  object 
 4   education_id      21402 non-null  int64  
 5   family_status     21402 non-null  object 
 6   family_status_id  21402 non-null  int64  
 7   gender            21402 non-null  object 
 8   income_type       21402 non-null  object 
 9   debt              21402 non-null  int64  
 10  total_income      21402 non-null  int32  
 11  purpose           21402 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 2.0+ MB


# 5. Обработка дубликатов 

Заменим неявные дубликаты в столбце `education` приведя их к нижнему регистру.

In [14]:
data["education"] = data["education"].str.lower()

Посчитаем явные дубликаты.

In [15]:
data.duplicated().sum()

71

Удалим явные дубликаты.

In [16]:
data = data.drop_duplicates()
data.duplicated().sum()

0

# 6. Категоризация данных 

Классифицируем заёмщиков по уровням доходов:

- 0–30000 — 'E'
- 30001–50000 — 'D'
- 50001–200000 — 'C'
- 200001–1000000 — 'B'
- 1000001 и выше — 'A'

In [17]:
def categorize_income(income):
    if income <= 30000:
        return 'E'
    elif income > 30000 and income <= 50000:
        return 'D'
    elif income > 50000 and income <= 200000:
        return 'C'
    elif income > 200000 and income <= 1000000:
        return 'B'
    else:
        return 'A'

data["total_income_category"] = data["total_income"].apply(categorize_income)
data["total_income_category"]

0        B
1        C
2        C
3        B
4        C
        ..
21520    B
21521    C
21522    C
21523    B
21524    C
Name: total_income_category, Length: 21331, dtype: object

Рассмотрим перечень уникальных целей взятия кредита из столбца `purpose`.

In [18]:
data["purpose"].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создайдим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- 'операции с автомобилем'
- 'операции с недвижимостью'
- 'проведение свадьбы'
- 'получение образования'

In [19]:
def categorize_purpose(purpose):
    if "авто" in purpose:
        return "операции с автомобилем"
    elif "жил" in purpose or "недвиж" in purpose:
        return "операции с недвижимостью"
    elif "свадьб" in purpose:
        return "проведение свадьбы"
    elif "образов" in purpose:
        return "получение образования"
    else:
        return None

data["purpose_category"] = data["purpose"].astype("string").apply(categorize_purpose)

# 7. Проверка гипетез

Сформулируем вопросы соответствующие проверяемым гипотезам:
- Есть ли зависимость между количеством детей `children` и возвратом кредита `debt` в срок?
- Есть ли зависимость между семейным положением `family_status_id` и возвратом кредита `debt` в срок ?
- Есть ли зависимость между уровнем дохода `total_income_category` и возвратом кредита `debt` в срок?
- Как разные цели кредита `purpose_category` влияют на его возврат `debt` в срок?

Для вывода о влиянии количества детей на просрочку долга посчитаем средние значения для заёмщиков с разным количеством детей.

In [20]:
pd.pivot_table(data, values="debt", index="children", aggfunc=["count", "mean"])

count      mean
           debt      debt
children                 
0         14091  0.075438
1          4808  0.092346
2          2052  0.094542
3           330  0.081818
4            41  0.097561
5             9  0.000000

<div class="alert alert-success">

**Артём:** 👍 Молодец, что используешь сводную таблицу.

</div>

Вывод: чем **больше детей** у заёмщика тем более он **склонен к просрочке**. Отклонения этой логики для 3 и 5 значений, вероятно, обусловлены меньшей выборкой таких заёмщиков. Вывод можно объяснить большим количеством незапланированных трат на детей, которые находятся на иждевении соответствующих заёмщиков.

<div class="alert alert-success">

**Артём:** 👍 Правильно, что смотришь на репрезентативность выборок.

</div>

In [21]:
100 - (data[data["children"] == 0]["debt"].mean() / data[data["children"] != 0]["debt"].mean()) * 100

18.359830438048476

Вероятность выплаты долга заёмщиком с детьми на 18% ниже, чем заёмщиком без детей

Для вывода о влиянии семейного положения на просрочку долга посчитаем средние значения для заёмщиков с разными семейными положениями.

In [22]:
family_status = pd.pivot_table(data, values="debt", index="family_status", aggfunc=["count", "mean"])
family_status

count      mean
                        debt      debt
family_status                         
Не женат / не замужем   2796  0.097639
в разводе               1189  0.070648
вдовец / вдова           951  0.066246
гражданский брак        4134  0.093130
женат / замужем        12261  0.075606

In [23]:
(family_status["mean"]["debt"].max() - family_status["mean"]["debt"].min()) * 100

3.1393428196206385

<div class="alert alert-success">

#### **Артём:** 👍 Лучше использовать переменные, а не абсолютные числа. На случай изменения датасета.

</div>

Можно сделать вывод, что в разных категориях заёмщиков по семейному положению разная вероятность выплаты долга, однако **влияние семейного положения незначительно** т.к. наибольшая разница между категориями достигает немногим более 3%.

Также можно заметить, что категории заёмщиков, которые были в отношениях и по какой-то причине перестали в них состоять немного более платёжеспособные. Для анализа причин этого замечания недостаточно информации.

<div class="alert alert-danger">
   
#### **Артём:** ❌ Во время исследования важно смотреть на контекст задачи.
    
Действительно несколько процентных пунктов - это не так много, но учитывая [уровень долговой нагрузки населения России перед банками](https://www.rbc.ru/finances/11/08/2021/6112dc5f9a79473a5acea094?), а это 23’900’000’000’000˙₽, то есть 2% это 478'000'000'000₽, так что лучше указать в финальном выводе.

</div>

Для вывода о влиянии класса доходов на просрочку долга посчитаем средние значения для заёмщиков с разными классами доходов.

In [24]:
pd.pivot_table(data, values="debt", index="total_income_category", aggfunc=["count", "mean"])

count      mean
                        debt      debt
total_income_category                 
A                         25  0.080000
B                       5014  0.070602
C                      15921  0.084982
D                        349  0.060172
E                         22  0.090909

In [25]:
data["debt"].mean()

0.08119638085415593

Сравнение платёжеспособности заёмщиков с разными классами доходов со средней платёжеспособностью заёмщиков позволяет сделать вывод, что заёмщики с **классом доходов E менее платёжеспособны**, чем остальные. При этом **самыми платёжеспособными являются заёмщики класса D, B**. 

Вероятно, информация недостаточно достоверная т.к. выборки для заёмщиков с калассами A, D, E значительно меньше остальных.

<div class="alert alert-success">

#### **Артём:** 👍 Дам тебе совет на будущее, как лучше проводить категоризацию количественных данных. **В этом проекте переделывать ничего не надо, так как заказчик уже предложил свои категории, его стоит проинформировать, что выборки крайне несбалансированны.** Если только как дополнительное исследование.
    
Сейчас самая маленькая категория в 728 раз меньше самой большой. Это может отразиться на результатах исследования. Когда мы делим на категории количественные данные, то лучше их делать примерно одинаковыми. Для этого можно воспользоваться методом `qcut()`. Пример:
    
```py
pd.qcut(df['total_income'], 4)
```
    
```
Name: total_income, Length: 21453, dtype: category
Categories (4, interval[float64]): [(20666.999, 107620.0]  (107620.0, 142594.0]  (142594.0, 195818.0]  (195818.0, 2265604.0]]
```
                                                                                                               
Что внутри: 
    
```py
pd.qcut(датафрейм['столбец'], количество_категорий)
``` 
    
Итого у нас 4 категории:
<ul>
  <li>меньше 107_620</li>
  <li>от 107_620 до 142_594</li>
  <li>от 142_594 до 195_818</li>
  <li>от 195_818</li>
    
Пример использования:
```py    
# Создаем категории дохода с помощью pd.qcut
data['total_qcut_category'] = pd.qcut(data['total_income'], 5, labels=['A', 'B', 'C', 'D', 'E'])

# Группируем данные по новому столбцу 'total_income_category'
pivot_table = data.groupby('total_qcut_category').agg(
    borrowers=('debt', 'size'),
    debtors=('debt', 'sum')
).reset_index()
```    
    
</ul>

Для вывода о влиянии цели займа на просрочку долга посчитаем средние значения для заёмщиков с разными целямт займа.

In [26]:
pd.pivot_table(data, values="debt", index="purpose_category", aggfunc=["count", "mean"])

count      mean
                           debt      debt
purpose_category                         
операции с автомобилем     4279  0.093480
операции с недвижимостью  10751  0.072551
получение образования      3988  0.092528
проведение свадьбы         2313  0.079118

In [27]:
(0.093480 - 0.072551) * 100

2.092899999999999

Разница между платёжеспособностью клиентов по целям займов незначительна и колеблется в пределах 2%.

Наиболее вероятны выплаты в срок по займам на операции с недвижимостью.

Наименнее вероятны выплаты в срок по займам на операции с автомобилем.

<div class="alert alert-success">

**Артём:** 👍 Правильно

</div>

[Оценка полноты данных](#gaps)

Т.к. количество пропусков в столбцах `days_employed` и `total_income` было одинаковым, целессообразно было проверить, не пропущены ли данные в одинаковых строках. Проверка показала, что пропуски были в одинаковых строках. Анализ корреляций для строк с пропусками показал, что пропуски случайны и не зависят от других атрибутов набора данных. 

Можно предположить, что информация о стаже и совокупных доходах была запрошена из соответствующих гос. органов и для этих заёмщиков отсутствует или была неверно запрошена. Стоит уточнить у создателей набора данных более точную информацию. 

<div class="alert alert-danger">
   
#### **Артём:** ❌ Получается, что причины пропущенных значений в 2174 одинаковых  строк, в разных столбцах были разными?
</div>

Данные в количественных переменных были заполнены медианными значениями соответствующих категорий с целью сохранения особенностей в категориях заёмщиков. Т.к. мы не знаем ничего о распределении данных в столбцах `days_employed` и `total_income`, заполним пропуски значениями медианы чтобы снизить искажение усреднения выборсами.

<div class="alert alert-danger">

#### **Артём:** ❌ А почему не средним? Чем медиана отличается от среднего?
    
</div>

# 8. Вывод

Наиболее **значимые различия** в факторе наличия и отсутствия детей. Заёмщики **без детей платёжеспособнее на 18%** по отношению к заёмщикам с детьми.

Другие категории влияют на платёжеспособность клиентов в меньшей степени.

**Наиболее платёжеспособны** заёмщики **с семейным статусом "в разводе" и "вдовец / вдова"**.
**Наименее платёжеспособны** заёмщики **с семейным статусом "не женат / не замужем" и состоящие "в гражданском браке"**.

Заметно, что категории заёмщиков, которые **были в отношениях** и по какой-то причине перестали в них состоять немного **более платёжеспособны**.

Сравнение платёжеспособности заёмщиков с разными классами доходов со средней платёжеспособностью заёмщиков позволяет сделать вывод, что заёмщики с **классом доходов E менее платёжеспособны**, чем остальные. При этом **самыми платёжеспособными являются заёмщики класса D, B**. 

Разница между платёжеспособностью клиентов по целям займов незначительна и колеблется в пределах 2%.
**Наиболее платёжеспособны** заёмщики **с целью операций с недвижимостью**.
**Наименее платёжеспособны** заёмщики **с целью операций с автомобилем**.

<div class="alert alert-success">

**Артём:** 👍 Хороший вывод. Аккуратно, последовательно приятно читать.

</div>

## 🏁 Итоги ревью
<a id="results"></a>
[В начало](#start)


<div class="alert alert-danger">

**Артём:** ❌ Yaroslav, спасибо за работу. Осталось исправить всего три комментария и можно идти к заказчику. Молодец! 👍
    
Для удобства, комментарии которые нужно скорректировать я добавил в меню **Contents** слева.
    
Жду новую версию проекта. 🙂

</div>

In [28]:
df1 = pd.DataFrame({'a': [1, 2, 3, 4], 'b': ['A', 'B', 'C', 'D']})
df2 = pd.DataFrame({'a': [2, 2, 2, 2], 'c': ['E', 'F', 'G', 'H']})
print(df1)
print()
print(df2)
print()
print (df1.join(df2, on='a', rsuffix='_y')['c']) 

   a  b
0  1  A
1  2  B
2  3  C
3  4  D

   a  c
0  2  E
1  2  F
2  2  G
3  2  H

0      F
1      G
2      H
3    NaN
Name: c, dtype: object
